src/Workspaces/Core/Portable/FindSymbols/FindReferences/FindReferenceCache.cs

```diff
@@ -56,186 +50,149 @@ public static SymbolInfo GetSymbolInfo(SemanticModel model, SyntaxNode node, Can
                 var initializers = GetConstructorInitializerTokens(syntaxFacts, root, cancellationToken);
-                Interlocked.CompareExchange(ref entry.ConstructorInitializerCache, initializers, null);
+                ImmutableInterlocked.InterlockedInitialize(ref entry.ConstructorInitializerCache, initializers);
             }
 
             return entry.ConstructorInitializerCache;
         }
 
-        private static List<SyntaxToken> GetConstructorInitializerTokens(
+        private static ImmutableArray<SyntaxToken> GetConstructorInitializerTokens(
             ISyntaxFactsService syntaxFacts, SyntaxNode root, CancellationToken cancellationToken)
         {
-            var initializers = new List<SyntaxToken>();
+            using var _ = ArrayBuilder<SyntaxToken>.GetInstance(out var initializers);
             foreach (var constructor in syntaxFacts.GetConstructors(root, cancellationToken))
             {
                 foreach (var token in constructor.DescendantTokens(descendIntoTrivia: false))
                 {
-                    if (!syntaxFacts.IsThisConstructorInitializer(token) && !syntaxFacts.IsBaseConstructorInitializer(token))
-                    {
-                        continue;
-                    }
-
-                    initializers.Add(token);
+                    if (syntaxFacts.IsThisConstructorInitializer(token) || syntaxFacts.IsBaseConstructorInitializer(token))
+                        initializers.Add(token);
                 }
             }
 
-            return initializers;
+            return initializers.ToImmutable();
         }
 
-        private static ConcurrentDictionary<SyntaxNode, SymbolInfo>? GetNodeCache(SemanticModel model)
+        private sealed class Entry
         {
-            var entry = GetCachedEntry(model);
-            return entry?.SymbolInfoCache;
-        }
-
-        private static Entry? GetCachedEntry(SemanticModel model)
-        {
-            using (s_gate.DisposableRead())
-            {
-                if (s_cache.TryGetValue(model, out var entry))
-                {
-                    return entry;
-                }
-
-                return null;
-            }
-        }
-
-        private static readonly Func<SemanticModel, Entry> s_entryCreator = _ => new Entry();
-
-        public static void Start(SemanticModel model)
-        {
-            Debug.Assert(model != null);
-
-            using (s_gate.DisposableWrite())
-            {
-                var entry = s_cache.GetOrAdd(model, s_entryCreator);
-                entry.Count++;
-            }
-        }
+            public ImmutableHashSet<string>? AliasNameSet;
+            public ImmutableArray<SyntaxToken> ConstructorInitializerCache;
 
-        public static void Stop(SemanticModel model)
-        {
-            if (model == null)
-            {
-                return;
-            }
+            public readonly ConcurrentDictionary<SyntaxNode, SymbolInfo> SymbolInfoCache = new();
```

Comment:
How big do these actually get in practice? Are these potentially holding thousands and thousands of items?

Reply:
so, in practice, this should just be the number of nodes *in* a document that we end up binding.  this will scale with the number of reasonable candidates we have for a particular symbol we're looking for.  e.g. if we're looking for 'Foo' this will likely be all the Foo identifier tokens within a file.  If there is a tiny amount, this will be tiny.  if there is a ton, this will be a lot.